external/kafka/src/main/scala/org/apache/gearpump/streaming/kafka/KafkaSource.scala

```diff
@@ -36,62 +39,131 @@ object KafkaSource {
+ * such that obsolete messages are dropped.
+ *
+ * @param config utility class for kafka config, see [[KafkaConfig]]
+ * @param timestampFilter filters out message based on timestamp
+ * @param fetchThread fetches messages and puts on a in-memory queue
+ * @param offsetManagers manages offset-to-timestamp storage for each [[TopicAndPartition]]
+ */
+class KafkaSource (config: KafkaConfig,
+                   timestampFilter: TimeStampFilter,
+                   private var fetchThread: Option[FetchThread] = None,
+                   private var offsetManagers: Map[TopicAndPartition, KafkaOffsetManager] = Map.empty[TopicAndPartition, KafkaOffsetManager])
 
-  private[kafka] def this(appName: String, config: KafkaConfig, grouper: KafkaGrouper, messageDecoder: MessageDecoder) =
-    this(appName, config, KafkaUtil.getTopicAndPartitions(KafkaUtil.connectZookeeper(config)(),
-      grouper, config.getConsumerTopics), messageDecoder)
+  extends TimeReplayableSource {
+  import org.apache.gearpump.streaming.kafka.KafkaSource._
 
-  def this(appName : String, taskId: TaskId, taskParallelism:Int, config: KafkaConfig, messageDecoder: MessageDecoder) =
-    this(appName: String, config, config.getGrouperFactory.getKafkaGrouper(taskId, taskParallelism), messageDecoder)
+  private var startTime: Option[TimeStamp] = None
+  private lazy val messageDecoder = config.getMessageDecoder
 
-  private var startTime: TimeStamp = 0L
+  /**
+   * @param topics comma-separated string of topics
+   * @param properties kafka consumer config
+   * @param timestampFilter filters out message based on timestamp
+   */
+  def this(topics: String, properties: Properties, timestampFilter: TimeStampFilter) = {
+    this(KafkaConfig.load("kafka.conf")
+      .withConsumerTopics(topics).withConsumerConfig(properties), timestampFilter)
+  }
 
-  LOG.debug(s"assigned ${offsetManagers.keySet}")
+  /**
+   * @param topics comma-separated string of topics
+   * @param zkConnect kafka consumer config `zookeeper.connect`
+   * @param messageDecoder decode gearpump [[Message]] from raw bytes
+   * @param timestampFilter filters out message based on timestamp
+   */
+  def this(topics: String, zkConnect: String, messageDecoder: String, timestampFilter: TimeStampFilter) = {
+    this(KafkaUtil.setZookeeperConnect(KafkaConfig.load("kafka.conf")
+        .withConsumerTopics(topics)
+        .withMessageDecoder(messageDecoder), zkConnect), timestampFilter)
+  }
 
   /**
-   * fetchThread will start from beginning (the earliest offset)
-   * on start offset not set
+   * @param topics comma-separated string of topics
+   * @param timestampFilter filters out message based on timestamp
    */
-  def startFromBeginning(): Unit = {
-    fetchThread.setDaemon(true)
-    fetchThread.start()
+  def this(topics: String, timestampFilter: TimeStampFilter) = {
+    this(KafkaConfig.load("kafka.conf").withConsumerTopics(topics), timestampFilter)
```

Comment:
suggest removing this constructor as it implicitly replying on kafka.conf to function. If there is no such file, it will fail with strange error. 

user must read this implementation to understand the behavior, which is not friendly.